build_utils/bump_changelog.py

```diff
@@ -0,0 +1,115 @@
+#! /usr/bin/env python
+
+"""Bump the current version number in CHANGELOG.rst
+
+Unreleased changes will be moved into a section for the new version number
+"""
+from argparse import ArgumentParser
+from dataclasses import dataclass
+from pathlib import Path
+import re
+
+from toolz.itertoolz import partition
+
+REPOSITORY_ADDRESS = "https://github.com/pace-neutrons/Euphonic"
+
+
+@dataclass
+class Block:
+    """CHANGELOG data corresponding to a single release"""
+    tag: str
+    previous_tag: str
+    content: str
+
+
+def parse_changelog(changelog_file: Path) -> list[Block]:
+    """Read all sections from changelog file
+
+    Each section should begin with two lines in format
+
+      `TAG <repository_address/compare/PREV_TAG...TAG>`_
+      --------------------------------------------------
+
+    except for 'Unreleased' which compares with HEAD
+    """
+
+    with open(changelog_file, "rt", encoding="utf8") as fd:
+        split_text = re.split(r"`(\S+) <\S+/compare/(\S+)\.\.\.\S+>`_\n-+", fd.read())
+
+    # First item is always empty?
+    split_text = split_text[1:]
+
+    # From Python 3.12 can replace partition with itertools.batched
+    blocks = [Block(*block_data) for block_data in partition(3, split_text)]
+
+    for block in blocks:
+        block.content = block.content.strip()
+
+    return blocks
+
+
+def bump_version(blocks: list[Block], tag: str) -> None:
+    """Create or update a new or existing block with content from "Unreleased" """
+
+    all_tags = [block.tag for block in blocks]
+
+    if all_tags[0] != "Unreleased":
+        raise ValueError("CHANGELOG should always begin with Unreleased")
+
+    if all_tags[1] != tag:
+        # This version is not in the CHANGELOG yet
+        previous_tag = all_tags[1]
+        blocks.insert(1, Block(tag, previous_tag, ""))
+
+    blocks[1].content = "\n\n".join(
+        [txt for txt in (blocks[0].content, blocks[1].content) if txt]
+    )
+
+    # Update "Unreleased" section
+    blocks[0].content = ""
+    blocks[0].previous_tag = tag
+
+
+def tag_to_header(tag: str, previous_tag: str) -> str:
+    """Produce header including github diff link"""
+    compare_tag = "HEAD" if tag == "Unreleased" else tag
+    header = f"`{tag} <{REPOSITORY_ADDRESS}/compare/{previous_tag}...{compare_tag}>`_"
```

Comment:
Wonder if it's worth making this act like a template str.
```python
LINK_STR = "`{tag} <{REPOSITORY_ADDRESS}/compare/{previous_tag}...{compare_tag}>`_"
...
split_text = re.split(re.escape(LINK_STR).format(tag=r"(\S+)", 
                                                 previous_tag=r"(\S+)", 
                                                 compare_tag=r"\S+") +
                      r"\n-+")
...
header = LINK_STR.format(tag=tag, previous_tag=previous_tag, compare_tag=compare_tag)
```

Reply:
Just realised the `re.escape` would escape the substitution brackets, which possibly makes this more complicated to handle than useful.

Reply:
Ooh, I'll give it a try

Reply:
I just had to replace the `re.escape` with a string replacement `. -> \.`.